File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>

typedef long long ll;

using namespace std;

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int N, M;
    cin >> N >> M;
    vector<int> arr;
    int max_val = 0;
    while (N--)
    {
        int num;
        cin >> num;
        max_val = max(num, max_val);
        arr.push_back(num);
    }

    ll left = 0, right = max_val;
    ll result = 0;
    while (left <= right)
    {
        ll mid = (left + right) / 2;
        ll total = 0;
        for (auto a : arr)
        {
            if (a >= mid)
            {
                total += a - mid;
            }
        }
        if (total >= M)
        {
            left = mid + 1;
            result = mid;
        }
        else
        {
            right = mid - 1;
        }
    }

    cout << result;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
if __name__ == '__main__':
    N, M = map(int, input().split())

    arr = list(map(int, input().split()))

    start = 0
    end = int(1e9)

    result = 0

    while start <= end:
        mid = start + end >> 1
        total = 0

        for a in arr:
            if a - mid >= 0:
                total += a - mid

        if total < M:
            end = mid - 1
        else:
            start = mid + 1
            result = mid

    print(result)